### ISSUE TITLE
NoopAnimationDriver isn't as Noop as its name suggests

### ISSUE THREAD

**Issue by sod** [angular/angular#36860]
# 🚀 Feature request

Reduce performance hit on ssr when using angular animations.

* angular 9.1.3

### Description

Profiling our ssr:
![image](https://user-images.githubusercontent.com/905328/80611510-f7e5f300-8a3a-11ea-8603-71297b2defba.png)

There is this `AnimationEngine.flush()` call at the end, that as far as I can tell, adds ~500 `ng-star-inserted` classes to all elements with an `*ngIf`. And also sprinked  some cpu cycles in between all over the app.

I found no official way to disable this behavior.

I monkeypatched these three methods away:

```typescript
const animations = require('@angular/animations/browser');
Object.assign(animations.ɵAnimationEngine.prototype, {onInsert: () => {}, register: () => {}, registerTrigger: () => {}});
```

Made the ssr ~5% faster, reduced memory consumption & reduced the html output by 8kb (by omitting all those ng-star-inserted classes). The html stil works as without the patch.

**Comment by pauldraper** [comment]
NoopAnimationDriver isn't no-op from a functional rendering perspective either: https://github.com/angular/angular/issues/27917 ngAfterViewChecked is fired with different DOM states if NoopAnimationDriver is present.

**Comment by matsko** [comment]
Hello @sod this is difficult to understand without a reproduction. I do understand that the NoOp driver does definitely do lots of extra work than is necessary. Could you put together a demo of some kind for us so we can investigate further?

**Comment by sod** [comment]
Reproduction https://github.com/sod/issue368860

## How to run:

```shell script
npm ci
ng run issue36860:server:production
node --inspect-brk dist/issue36860/server/main.js
```

1. Open node debugger in chrome
2. Start flame chart profiler
3. Open http://localhost:4000
4. Inspect profile

## How to interpret
Generates this flame chart, where a good chunk of the time spent is in AnimationEngine.flush, that adds `ng-star-inserted` css class to 1000 elements (the demo creates 1000 elements via *ngIf)
- https://user-images.githubusercontent.com/905328/83083525-a30fb980-a086-11ea-96d7-8ea777619941.png

Here the cpu profile generated in prod mode as direct download
- https://github.com/sod/issue368860/blob/master/CPU-profile-issue36860.cpuprofile

**Comment by sod** [comment]
We now ditched @angular/animations/browser now in favor of native css animations. Reduced the minified build by 80kb and made the page & ssr faster by default. And as a bonus it works properly in safari (was the biggest reason the remove this dependency)

So I don't need this fixed anymore.

Who is interested in doing the same, I wrote [ClassActiveAnimateDirective to help with elements entering the DOM](https://gist.github.com/sod/26431ae994311e226829d57b7da827c9) and [ObserveRemovalsPipe for elements leaving the DOM in an ngFor](https://gist.github.com/sod/4d3aa97046e31cf4bc83c41463735fce). Who stil uses NgZone has to wrap all zone timeouts in runOutsideAngular of course.
